apps/aecore/src/aec_pinning_agent.erl

```diff
@@ -0,0 +1,168 @@
+          %% parameters, namely:
+          %% * parent chain's protocol version
+          %% * expected min_gas on both miner and protocol level
+          %% * current gas prices
+          fee          => Fee,
+          nonce        => Nonce,
+          payload      => PinPayload},
+    {ok, SpendTx} = aec_spend_tx:new(TxArgs),
+    SpendTx.
+
+post_pin_tx(SignedSpendTx, NodeSpec) ->
+    Transaction = aeser_api_encoder:encode(transaction, aetx_sign:serialize_to_binary(SignedSpendTx)),
+    Body = #{<<"tx">> => Transaction},
+    Path = <<"/v3/transactions">>,
+    post_request(Path, Body, NodeSpec, 5000).
+
+%% TODO This function copied from aec_test_utils as that module is not available
+%% in normal builds
+%% TODO - wallet interaction of some kind to get privKey
+%% This is a horrible hack for now
+%% Maybe aec_preset_keys??
+%% {ok, Sig} = aec_preset_keys:sign_tx(SpendTx, AccountId)
+%% TODO: redo this whole thing
+-define(VALID_PRIVK(K), byte_size(K) =:= 64).
+
+get_request(Path, NodeSpec, Timeout) ->
+    try
+    Url = aehttpc:url(NodeSpec),
+    UrlPath = lists:concat([Url, binary_to_list(Path)]),
+    Req = {UrlPath, []},
+    HTTPOpt = [{timeout, Timeout}],
+    Opt = [],
+    %% lager:debug("Req: ~p, with URL: ~ts", [Req, Url]),
+    case httpc:request(get, Req, HTTPOpt, Opt) of
+        {ok, {{_, 200 = _Code, _}, _, Res}} ->
+            %% lager:debug("Req: ~p, Res: ~p with URL: ~ts", [Req, Res, Url]),
+            {ok, jsx:decode(list_to_binary(Res), [return_maps])};
+        {ok, {{_, 404 = _Code, _}, _, "{\"reason\":\"Block not found\"}"}} ->
+            {error, not_found};
+        {ok, {{_, Code, _}, _, Res}} when Code >= 400 ->
+            lager:debug("Req: ~p, Res: ~p with URL: ~ts, Code ~p", [Req, Res, Url, Code]),
+            {error, not_found};
+        {error, Err} ->
+            lager:debug("Req: ~p, ERROR: ~p with URL: ~ts", [Req, Err, Url]),
+            {error, not_found}
+    end
+  catch E:R:S ->
+    lager:info("Error: ~p Reason: ~p Stacktrace: ~p", [E, R, S]),
+    {error, {E, R, S}}
+  end.
+
+-spec post_request(binary(), map(), aehttpc:node_spec(), integer()) -> {ok, map()} | {error, term()}.
+post_request(Path, Body, NodeSpec, Timeout) ->
+  try
+    Url = aehttpc:url(NodeSpec),
+    UrlPath = lists:concat([Url, binary_to_list(Path)]),
+    Req = {UrlPath, [], "application/json", jsx:encode(Body)},
+    HTTPOpt = [{timeout, Timeout}],
+    Opt = [],
+    case httpc:request(post, Req, HTTPOpt, Opt) of
+        {ok, {{_, 200 = _Code, _}, _, Res}} ->
+            lager:debug("Req: ~p, Res: ~p with URL: ~ts", [Req, Res, Url]),
+            {ok, jsx:decode(list_to_binary(Res), [return_maps])};
+        {ok, {{_ ,400, _}, _, Res}} ->
+            lager:debug("Req: ~p, Res: ~p with URL: ~ts", [Req, Res, Url]),
+            {error, 400, jsx:decode(list_to_binary(Res), [return_maps])}
+    end
+  catch E:R:S ->
+    lager:info("Error: ~p Reason: ~p Stacktrace: ~p", [E, R, S]),
+    {error, {E, R, S}}
+  end.
```

Comment:
Switching to indentation 2 (should be 4) in the middle of the file?

Reply:
Sure thing.